sound/soc/sof/ipc4-pcm.c

```diff
@@ -276,6 +283,36 @@ static int sof_ipc4_chain_dma_trigger(struct snd_sof_dev *sdev,
 	return sof_ipc_tx_message_no_reply(sdev->ipc, &msg, 0);
 }
 
+static void sof_ipc4_sort_pipeline(struct ipc4_pipeline_set_state_data *trigger_list,
+				   u8 *pipe_priority, int state)
+{
+	int i, j;
+
+	for (i = 1; i < trigger_list->count - 1; i++) {
+		for (j = 0; j < trigger_list->count - i; j++) {
+			bool swap_condition;
+
+			/* high priority will be sent first */
+			if (state == SOF_IPC4_PIPE_RUNNING || state == SOF_IPC4_PIPE_RESET)
+				swap_condition = pipe_priority[j] < pipe_priority[j + 1] ? 1 : 0;
+			else
+				swap_condition = pipe_priority[j] > pipe_priority[j + 1] ? 1 : 0;
+
+			if (swap_condition) {
+				int tmp;
+
+				tmp = pipe_priority[j];
+				pipe_priority[j] = pipe_priority[j + 1];
+				pipe_priority[j + 1] = tmp;
+
+				tmp = trigger_list->pipeline_instance_ids[j];
+				trigger_list->pipeline_instance_ids[j] =
+					trigger_list->pipeline_instance_ids[j + 1];
+				trigger_list->pipeline_instance_ids[j + 1] = tmp;
+			}
+		}
+	}
```

Comment:
`sort_r()` ?

Reply:
I did investigate sort() function in Linux. It is much more complex for this simple sort and we have to pack priority and pipeline id in a array to match up with sort interface. 
````c
void sort(void *base, size_t num, size_t size,
          int (*cmp_func)(const void *, const void *),
          void (*swap_func)(void *, void *, int size))
````

Reply:
code changed